File: src/main/java/de/bypixeltv/redivelocity/commands/RediVelocityCommand.java
/*
 * Copyright (c) 2025.
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package de.bypixeltv.redivelocity.commands;

import com.velocitypowered.api.proxy.ProxyServer;
import com.velocitypowered.api.proxy.server.ServerPing;
import de.bypixeltv.redivelocity.RediVelocityLogger;
import de.bypixeltv.redivelocity.config.ConfigLoader;
import de.bypixeltv.redivelocity.jedisWrapper.RedisController;
import dev.jorel.commandapi.CommandAPICommand;
import dev.jorel.commandapi.arguments.ArgumentSuggestions;
import dev.jorel.commandapi.arguments.StringArgument;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import net.kyori.adventure.text.minimessage.MiniMessage;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

@Singleton
public class RediVelocityCommand {

    private final String prefix;
    private final MiniMessage miniMessage;
    private final RedisController redisController;
    private final ProxyServer proxy;

    @Inject
    public RediVelocityCommand(RedisController redisController, ProxyServer proxy, RediVelocityLogger rediVelocityLogger) {
        ConfigLoader configLoader = new ConfigLoader(rediVelocityLogger);
        configLoader.load();
        this.prefix = configLoader.getConfig().getMessages().getPrefix();
        this.miniMessage = MiniMessage.miniMessage();
        this.redisController = redisController;
        this.proxy = proxy;
    }

    public void register() {
        new CommandAPICommand("redivelocity")
                .withAliases("rv", "rediv")
                .withSubcommands(
                        createPlayerSubcommands(),
                        createProxySubcommands()
                ).register();
    }

    private CommandAPICommand createPlayerSubcommands() {
        return new CommandAPICommand("player")
                .withSubcommands(
                        createPlayerProxyCommand(),
                        createPlayerLastSeenCommand(),
                        createPlayerIpCommand(),
                        createPlayerUuidCommand(),
                        createPlayerServerCommand(),
                        createPlayerServersCommand()
                );
    }

    private CommandAPICommand createPlayerProxyCommand() {
        return new CommandAPICommand("proxy")
                .withArguments(new StringArgument("player").replaceSuggestions(ArgumentSuggestions.stringCollection(_ -> redisController.getAllHashValues("rv-players-name"))))
                .withPermission("redivelocity.admin.player.proxy")
                .executes((sender, args) -> {
                    String playerName = (String) args.get(0);
                    String playerKey = redisController.getHashKeyByValue("rv-players-name", playerName);
                    if (playerKey == null) {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> does not exist.</gray>"));
                        return;
                    }
                    String playerProxy = redisController.getHashField("rv-players-proxy", playerKey);
                    if (playerProxy != null) {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> is connected to proxy: <aqua>" + playerProxy + "</aqua></gray>"));
                    } else {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> is <red>offline</red>.</gray>"));
                    }
                });
    }

    private CommandAPICommand createPlayerLastSeenCommand() {
        return new CommandAPICommand("lastseen")
                .withArguments(new StringArgument("player").replaceSuggestions(ArgumentSuggestions.stringCollection(_ -> redisController.getAllHashValues("rv-players-name"))))
                .withPermission("redivelocity.admin.player.lastseen")
                .executes((sender, args) -> {
                    String playerName = (String) args.get(0);
                    String playerKey = redisController.getHashKeyByValue("rv-players-name", playerName);
                    if (playerKey == null) {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> does not exist.</gray>"));
                        return;
                    }
                    String lastSeen = redisController.getHashField("rv-players-lastseen", playerKey);
                    assert playerName != null;
                    boolean isOnline = redisController.getHashField("rv-players-name", playerKey).contains(playerName);
                    if (!isOnline) {
                        if (lastSeen != null) {
                            sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> was last seen <aqua>" + lastSeen + "</aqua>.</gray>"));
                        } else {
                            sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> was never seen before.</gray>"));
                        }
                    } else {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> is currently <green>online</green>.</gray>"));
                    }
                });
    }

    private CommandAPICommand createPlayerIpCommand() {
        return new CommandAPICommand("ip")
                .withArguments(new StringArgument("player").replaceSuggestions(ArgumentSuggestions.stringCollection(_ -> redisController.getAllHashValues("rv-players-name"))))
                .withPermission("redivelocity.admin.player.ip")
                .executes((sender, args) -> {
                    String playerName = (String) args.get(0);
                    String playerKey = redisController.getHashKeyByValue("rv-players-name", playerName);
                    if (playerKey == null) {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> does not exist.</gray>"));
                        return;
                    }
                    String playerIp = redisController.getHashField("rv-players-ip", playerKey);
                    if (playerIp != null) {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> is connected with IP: <aqua><hover:show_text:'<aqua>Click to copy</aqua>'><click:copy_to_clipboard:" + playerIp + ">" + playerIp + "</click></hover></aqua></gray>"));
                    } else {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> is <red>offline</red>.</gray>"));
                    }
                });
    }

    private CommandAPICommand createPlayerUuidCommand() {
        return new CommandAPICommand("uuid")
                .withArguments(new StringArgument("player").replaceSuggestions(ArgumentSuggestions.stringCollection(_ -> redisController.getAllHashValues("rv-players-name"))))
                .withPermission("redivelocity.admin.player.uuid")
                .executes((sender, args) -> {
                    String playerName = (String) args.get(0);
                    String playerUuid = redisController.getHashKeyByValue("rv-players-name", playerName);
                    if (playerUuid != null) {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> has the UUID: <aqua><hover:show_text:'<aqua>Click to copy</aqua>'><click:copy_to_clipboard:" + playerUuid + ">" + playerUuid + "</click></hover></aqua></gray>"));
                    } else {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> does not exist.</gray>"));
                    }
                });
    }

    private CommandAPICommand createPlayerServerCommand() {
        return new CommandAPICommand("server")
                .withArguments(new StringArgument("player").replaceSuggestions(ArgumentSuggestions.stringCollection(_ -> redisController.getAllHashValues("rv-players-name"))))
                .withPermission("redivelocity.admin.player.uuid")
                .executes((sender, args) -> {
                    String playerName = (String) args.get(0);
                    String playerKey = redisController.getHashKeyByValue("rv-players-name", playerName);
                    if (playerKey == null) {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> does not exist.</gray>"));
                        return;
                    }
                    String playerServer = redisController.getHashField("rv-players-server", playerKey);
                    if (playerServer != null) {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</aqua> is currently on server: <aqua>" + playerServer + "</aqua></gray>"));
                    } else {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>The player <aqua>" + playerName + "</a> is <red>offline</red>.</gray>"));
                    }
                });
    }

    private CommandAPICommand createPlayerServersCommand() {
        return new CommandAPICommand("servers")
                .withPermission("redivelocity.admin.player.servers")
                .executes((sender, _) -> {
                    Map<String, String> players = redisController.getHashValuesAsPair("rv-players-server");
                    List<String> playersPrettyNames = players.entrySet().stream()
                            .map(entry -> prefix + " <aqua>" + redisController.getHashField("rv-players-name", entry.getKey()) + "</aqua> <dark_gray>(<grey>Server: <aqua>" + entry.getValue() + "</aqua></grey>)</dark_gray>")
                            .collect(Collectors.toList());
                    String playersPrettyString = String.join("<br>", playersPrettyNames);
                    if (playersPrettyNames.isEmpty()) {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>There are currently no players online. (Stupid msg, but when bug, pls report or fix it)</gray>"));
                    } else {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>Current servers of all online players:<br>" + playersPrettyString + "</gray>"));
                    }
                });
    }

    private CommandAPICommand createProxySubcommands() {
        return new CommandAPICommand("proxy")
                .withSubcommands(
                        createProxyListCommand(),
                        createProxyPlayersCommand(),
                        createProxyPlayerCountCommand(),
                        createProxyServersCommand()
                );
    }

    private CommandAPICommand createProxyListCommand() {
        return new CommandAPICommand("list")
                .withPermission("redivelocity.admin.proxy.list")
                .executes((sender, _) -> {
                    Set<String> proxies = redisController.getAllHashFields("rv-proxies");
                    List<String> proxiesPrettyNames = proxies.stream()
                            .map(proxyId -> prefix + " <aqua>" + proxyId + "</aqua> <dark_grey>(<grey>Players: </grey><aqua>" + redisController.getHashField("rv-proxy-players", proxyId) + "</aqua>)</dark_grey>")
                            .collect(Collectors.toList());
                    String proxiesPrettyString = String.join("<br>", proxiesPrettyNames);
                    if (proxies.isEmpty()) {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>There are currently no connected proxies.</gray>"));
                    } else {
                        sender.sendMessage(miniMessage.deserialize(prefix + " <gray>Currently connected proxies:<br>" + proxiesPrettyString + "</gray>"));
                    }
                });
    }

    private CommandAPICommand createProxyPlayersCommand() {
        return new CommandAPICommand("players")
                .withOptionalArguments(new StringArgument("proxy").replaceSuggestions(ArgumentSuggestions.stringCollection(_ -> redisController.getAllHashFields("rv-proxies"))))
                .withPermission("redivelocity.admin.proxy.players")
                .executes((sender, args) -> {
                    String proxyId = (String) args.getOptional(0).orElse(null);
                    List<String> players = redisController.getAllHashValues("rv-players-name");
                    List<String> playersPrettyNames = players.stream()
                            .map(player -> {
                                String playerProxy = redisController.getHashField("rv-players-proxy", redisController.getHashKeyByValue("rv-players-name", player));
                                if (proxyId == null) {
                                    return prefix + " <aqua>" + player + "</aqua> <dark_gray>(<aqua>" + playerProxy + "</aqua>)</dark_gray>";
                                } else if (playerProxy.equals(proxyId)) {
                                    return prefix + " <aqua>" + player + "</aqua>";
                                }
                                return null;
                            })
                            .filter(Objects::nonNull)
                            .collect(Collectors.toList());
                    String playersPrettyString = String.join("<br>", playersPrettyNames);
                    if (playersPrettyNames.isEmpty()) {
                        if (proxyId == null) {
                            sender.sendMessage(miniMessage.deserialize(prefix + " <gray>There are currently no players online.</gray>"));
                        } else {
                            sender.sendMessage(miniMessage.deserialize(prefix + " <gray>There are currently no players online on proxy <aqua>" + proxyId + "</aqua>.</gray>"));
                        }
                    } else {
                        if (proxyId != null) {
                            sender.sendMessage(miniMessage.deserialize(prefix + " <gray>Currently online players on proxy <aqua>" + proxyId + "</aqua>:<br>" + playersPrettyString + "</gray>"));
                        } else {
                            sender.sendMessage(miniMessage.deserialize(prefix + " <gray>Currently online players:<br>" + playersPrettyString + "</gray>"));
                        }
                    }
                });
    }

    private CommandAPICommand createProxyPlayerCountCommand() {
        return new CommandAPICommand("playercount")
                .withOptionalArguments(new StringArgument("proxy").replaceSuggestions(ArgumentSuggestions.stringCollection(_ -> redisController.getAllHashFields("rv-proxies"))))
                .withPermission("redivelocity.admin.proxy.playercount")
                .executes((sender, args) -> {
                    String proxyId = (String) args.getOptional(0).orElse(null);
                    if (proxyId == null) {
                        String playerCount = redisController.getString("rv-global-playercount");
                        if (playerCount != null) {
                            sender.sendMessage(miniMessage.deserialize(prefix + " <gray>There are currently <aqua>" + playerCount + "</aqua> players online.</gray>"));
                        } else {
                            sender.sendMessage(miniMessage.deserialize(prefix + " <gray>There are currently no players online.</gray>"));
                        }
                    } else {
                        String playerCount = redisController.getHashField("rv-proxy-players", proxyId);
                        if (playerCount != null) {
                            sender.sendMessage(miniMessage.deserialize(prefix + " <gray>There are currently <aqua>" + playerCount + "</aqua> players online on proxy <aqua>" + proxyId + "</aqua>.</gray>"));
                        } else {
                            sender.sendMessage(miniMessage.deserialize(prefix + " <gray>There are currently no players online on proxy <aqua>" + proxyId + "</aqua>.</gray>"));
                        }
                    }
                });
    }

    private CommandAPICommand createProxyServersCommand() {
        return new CommandAPICommand("servers")
                .withPermission("redivelocity.admin.proxy.servers")
                .executes((sender, _) -> {
                    List<CompletableFuture<String>> futures = proxy.getAllServers().stream()
                            .map(server -> CompletableFuture.supplyAsync(() -> {
                                try {
                                    ServerPing result = server.ping().get();
                                    return prefix + " <color:#0dbf00>●</color> <aqua>" + server.getServerInfo().getName() + "</aqua> <dark_gray>(<grey>Address: <aqua>" + server.getServerInfo().getAddress() + "</aqua>, Version: <aqua>" + result.getVersion().getProtocol() + ", " + result.getVersion().getName() + "</aqua></grey>)</dark_gray>";
                                } catch (Exception e) {
                                    return prefix + " <color:#f00000>●</color> <aqua>" + server.getServerInfo().getName() + "</aqua> <dark_gray>(<grey>Address: <aqua>" + server.getServerInfo().getAddress() + "</aqua></grey>)</dark_gray>";
                                }
                            }))
                            .toList();
                    CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).thenRun(() -> {
                        List<String> proxyRegisteredServersPrettyNames = futures.stream()
                                .map(CompletableFuture::join)
                                .collect(Collectors.toList());
                        String proxyRegisteredServersPrettyString = String.join("<br>", proxyRegisteredServersPrettyNames);
                        if (!proxyRegisteredServersPrettyNames.isEmpty()) {
                            sender.sendMessage(miniMessage.deserialize(prefix + " <gray>Currently registered servers:<br>" + proxyRegisteredServersPrettyString + "</gray>"));
                        } else {
                            sender.sendMessage(miniMessage.deserialize(prefix + " <gray>There are currently no registered servers.</gray>"));
                        }
                    });
                });
    }
}

File: src/main/kotlin/de/bypixeltv/redivelocity/commands/RediVelocityCommand.kt
package de.bypixeltv.redivelocity.commands

import com.velocitypowered.api.proxy.ProxyServer
import de.bypixeltv.redivelocity.RediVelocity
import de.bypixeltv.redivelocity.config.ConfigLoader
import de.bypixeltv.redivelocity.utils.MojangUtils
import de.bypixeltv.redivelocity.utils.asDateString
import dev.jorel.commandapi.CommandAPICommand
import dev.jorel.commandapi.arguments.ArgumentSuggestions
import dev.jorel.commandapi.arguments.StringArgument
import dev.jorel.commandapi.executors.CommandExecutor
import jakarta.inject.Inject
import jakarta.inject.Provider
import jakarta.inject.Singleton
import kotlinx.coroutines.*
import net.kyori.adventure.text.minimessage.MiniMessage
import java.util.*
import kotlin.jvm.optionals.getOrNull

@Singleton
class RediVelocityCommand @Inject constructor(
    rediVelocityProvider: Provider<RediVelocity>,
    private val proxy: ProxyServer,
    private val mojangUtilsProvider: Provider<MojangUtils>
) {

    private val configLoader: ConfigLoader = ConfigLoader("plugins/redivelocity/config.yml").apply { load() }

    private val config = configLoader.config

    private val miniMessage = MiniMessage.miniMessage()
    private val prefix = config?.messages?.prefix

    private val redisController = rediVelocityProvider.get().getRedisController()

    private val mojangUtils: MojangUtils
        get() = mojangUtilsProvider.get()

    @OptIn(DelicateCoroutinesApi::class)
    @Suppress("UNUSED")
    fun register() {
        CommandAPICommand("redivelocity")
            .withAliases("rv", "rediv", "redisvelocity", "redisv")
            .withPermission("redivelocity.admin")
            .withSubcommands(
                CommandAPICommand("player")
                    .withSubcommands(
                        CommandAPICommand("proxy")
                            .withArguments(StringArgument("playerProxy").replaceSuggestions(ArgumentSuggestions.stringCollection {
                                redisController.getAllHashValues("rv-players-name")
                            }))
                            .withPermission("redivelocity.admin.player.proxy")
                            .executes(CommandExecutor { sender, args ->
                                val playerName = args[0] as String
                                val playerProxy = redisController.getHashField("rv-players-proxy",
                                    mojangUtils.getUUID(playerName).toString()
                                )
                                if (playerProxy != null) {
                                    sender.sendMessage(miniMessage.deserialize("$prefix <gray>The player <aqua>$playerName</aqua> is connected to proxy: <aqua>$playerProxy</aqua></gray>"))
                                } else {
                                    sender.sendMessage(miniMessage.deserialize("$prefix <gray>The player <aqua>$playerName</aqua> is <red>offline</red>.</gray>"))
                                }
                            }),
                        CommandAPICommand("lastseen")
                            .withArguments(StringArgument("playerLastSeen").replaceSuggestions(ArgumentSuggestions.stringCollection {
                                redisController.getAllHashValues("rv-players-name")
                            }))
                            .withPermission("redivelocity.admin.player.lastseen")
                            .executes(CommandExecutor { sender, args ->
                                val playerName = args[0] as String
                                val lastSeen = redisController.getHashField("rv-players-lastseen", mojangUtils.getUUID(playerName).toString())
                                val isOnline: Boolean =
                                    redisController.getHashField("rv-players-name", mojangUtils.getUUID(playerName).toString())?.contains(playerName) == true
                                if (!isOnline) {
                                    if (lastSeen != null) {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>The player <aqua>$playerName</aqua> was last seen <aqua>${lastSeen.toLong().asDateString()}</aqua>.</gray>"))
                                    } else {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>The player <aqua>$playerName</aqua> was never seen before.</gray>"))
                                    }
                                } else {
                                    sender.sendMessage(miniMessage.deserialize("$prefix <gray>The player <aqua>$playerName</aqua> is currently <green>online</green>.</gray>"))
                                }
                            }),
                        CommandAPICommand("ip")
                            .withArguments(StringArgument("playerIp").replaceSuggestions(ArgumentSuggestions.stringCollection {
                                redisController.getAllHashValues("rv-players-name")
                            }))
                            .withPermission("redivelocity.admin.player.ip")
                            .executes(CommandExecutor { sender, args ->
                                val playerName = args[0] as String
                                val playerIp = redisController.getHashField("rv-players-ip", mojangUtils.getUUID(playerName).toString())
                                if (playerIp != null) {
                                    sender.sendMessage(miniMessage.deserialize("$prefix <gray>The player <aqua>$playerName</aqua> is connected with IP: <aqua><hover:show_text:'<aqua>Click to copy</aqua>'><click:copy_to_clipboard:$playerIp>$playerIp</click></hover></aqua></gray>"))
                                } else {
                                    sender.sendMessage(miniMessage.deserialize("$prefix <gray>The player <aqua>$playerName</aqua> is <red>offline</red>.</gray>"))
                                }
                            }),
                        CommandAPICommand("uuid")
                            .withArguments(StringArgument("playerUuid").replaceSuggestions(ArgumentSuggestions.stringCollection {
                                redisController.getAllHashValues("rv-players-name")
                            }))
                            .withPermission("redivelocity.admin.player.uuid")
                            .executes(CommandExecutor { sender, args ->
                                val playerName = args[0] as String
                                val playerUuid = mojangUtils.getUUID(playerName).toString()
                                sender.sendMessage(miniMessage.deserialize("$prefix <gray>The player <aqua>$playerName</aqua> has the UUID: <aqua><hover:show_text:'<aqua>Click to copy</aqua>'><click:copy_to_clipboard:$playerUuid>$playerUuid</click></hover></gray>"))
                            }),
                        CommandAPICommand("server")
                            .withArguments(StringArgument("playerUuid").replaceSuggestions(ArgumentSuggestions.stringCollection {
                                redisController.getAllHashValues("rv-players-name")
                            }))
                            .withPermission("redivelocity.admin.player.server")
                            .executes(CommandExecutor { sender, args ->
                                val playerName = args[0] as String
                                val playerServer = redisController.getHashField("rv-players-server", mojangUtils.getUUID(playerName).toString())
                                if (playerServer != null) {
                                    sender.sendMessage(miniMessage.deserialize("$prefix <gray>The player <aqua>$playerName</aqua> is connected to server: <aqua>$playerServer</aqua></gray>"))
                                } else {
                                    sender.sendMessage(miniMessage.deserialize("$prefix <gray>The player <aqua>$playerName</aqua> is  currently <red>not</red> connected to a server.</gray>"))
                                }
                            })
                    ),
                CommandAPICommand("proxy")
                    .withSubcommands(
                        CommandAPICommand("list")
                            .withPermission("redivelocity.admin.proxy.list")
                            .executes(CommandExecutor { sender, _ ->
                                val proxies = redisController.getList("rv-proxies")
                                val proxiesPrettyNames: MutableList<String> = mutableListOf()
                                proxies?.forEach { proxyId ->
                                    proxiesPrettyNames.add("$prefix <aqua>$proxyId</aqua> <dark_grey>(<grey>Players: </grey><aqua>${redisController.getHashField("rv-proxy-players", proxyId)}</aqua>)</dark_grey>")
                                }
                                val proxiesPrettyString = proxiesPrettyNames.joinToString(separator = "<br>")
                                if (proxies != null) {
                                    if (proxies.isEmpty()) {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>There are currently no connected proxies.</gray>"))
                                        return@CommandExecutor
                                    } else {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>Currently connected proxies:<br>$proxiesPrettyString</gray>"))
                                    }
                                } else {
                                    sender.sendMessage(miniMessage.deserialize("$prefix <gray>There are currently no connected proxies.</gray>"))
                                    return@CommandExecutor
                                }
                            }),
                        CommandAPICommand("players")
                            .withOptionalArguments(StringArgument("proxy").replaceSuggestions(ArgumentSuggestions.stringCollection {
                                redisController.getList("rv-proxies")
                            }))
                            .withPermission("redivelocity.admin.proxy.players")
                            .executes(CommandExecutor { sender, args ->
                                val proxyId = args.getOptional(0).getOrNull() as? String
                                val players = redisController.getAllHashValues("rv-players-name")
                                val playersPrettyNames: MutableList<String> = mutableListOf()
                                val playerNames: MutableList<String> = mutableListOf()
                                players?.forEach { player ->
                                    val playerProxy = redisController.getHashField("rv-players-proxy", mojangUtils.getUUID(player).toString())
                                    if (proxyId == null) {
                                        playerNames.add(player)
                                        playersPrettyNames.add("$prefix <aqua>$player</aqua> <dark_gray>(<aqua>$playerProxy</aqua>)</dark_gray>")
                                    } else {
                                        if (playerProxy == proxyId) {
                                            playerNames.add(player)
                                            playersPrettyNames.add("$prefix <aqua>$player</aqua>")
                                        }
                                    }
                                }
                                val playersPrettyString = playersPrettyNames.joinToString(separator = "<br>")
                                if (playerNames.isEmpty()) {
                                    if (proxyId == null) {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>There are currently no players online.</gray>"))
                                        return@CommandExecutor
                                    } else {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>There are currently no players online on proxy <aqua>$proxyId</aqua>.</gray>"))
                                        return@CommandExecutor
                                    }
                                } else {
                                    if (proxyId != null) {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>Currently online players on proxy <aqua>$proxyId</aqua>:<br>$playersPrettyString</gray>"))
                                    } else {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>Currently online players:<br>$playersPrettyString</gray>"))

                                    }
                                }
                            }),
                        CommandAPICommand("playercount")
                            .withOptionalArguments(StringArgument("proxy").replaceSuggestions(ArgumentSuggestions.stringCollection {
                                redisController.getList("rv-proxies")
                            }))
                            .withPermission("redivelocity.admin.proxy.playercount")
                            .executes(CommandExecutor { sender, args ->
                                val proxyId = args.getOptional(0).getOrNull() as? String
                                if (proxyId == null) {
                                    val playerCount = redisController.getString("rv-global-playercount")
                                    if (playerCount != null) {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>There are currently <aqua>$playerCount</aqua> players online.</gray>"))
                                    } else {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>There are currently no players online.</gray>"))
                                    }
                                } else {
                                    val playerCount = redisController.getHashField("rv-proxy-players", proxyId)
                                    if (playerCount != null) {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>There are currently <aqua>$playerCount</aqua> players online on proxy <aqua>$proxyId</aqua>.</gray>"))
                                    } else {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>There are currently no players online on proxy <aqua>$proxyId</aqua>.</gray>"))
                                    }
                                }
                            }),
                        CommandAPICommand("servers")
                            .withPermission("redivelocity.admin.proxy.servers")
                            .executes(CommandExecutor { sender, _ ->
                                GlobalScope.launch {
                                    val proxyRegisteredServers = proxy.allServers
                                    val futures = proxyRegisteredServers.map { server ->
                                        async {
                                            try {
                                                val result = server.ping().get()
                                                "$prefix <color:#0dbf00>●</color> <aqua>${server.serverInfo.name}</aqua> <dark_gray>(<grey>Address: <aqua>${server.serverInfo.address}</aqua>, Playercount: <aqua>${server.playersConnected.size}</aqua>, Version: <aqua>${result.version.protocol}, ${result.version.name}</aqua></grey>)</dark_gray>"
                                            } catch (e: Exception) {
                                                "$prefix <color:#f00000>●</color> <aqua>${server.serverInfo.name}</aqua> <dark_gray>(<grey>Address: <aqua>${server.serverInfo.address}</aqua></grey>)</dark_gray>"
                                            }
                                        }
                                    }
                                    val proxyRegisteredServersPrettyNames = futures.awaitAll()
                                    val proxyRegisteredServersPrettyString = proxyRegisteredServersPrettyNames.joinToString(separator = "<br>")
                                    if (proxyRegisteredServers.isNotEmpty()) {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>Currently registered servers:<br>$proxyRegisteredServersPrettyString</gray>"))
                                    } else {
                                        sender.sendMessage(miniMessage.deserialize("$prefix <gray>There are currently no registered servers.</gray>"))
                                    }
                                }
                            })
                    ),
                CommandAPICommand("blacklist")
                    .withSubcommands(
                        CommandAPICommand("add")
                            .withArguments(StringArgument("proxyBlacklistAdd").replaceSuggestions(ArgumentSuggestions.stringCollection {
                                redisController.getAllHashValues("rv-players-name")
                            }))
                            .withPermission("redivelocity.admin.proxy.blacklist.add")
                            .executes(CommandExecutor { sender, args ->
                                val playerName = args[0] as String
                                val playerUUID = mojangUtils.getUUID(playerName).toString()
                                if (!mojangUtils.isValidUUID(playerUUID)) {
                                    sender.sendMessage(miniMessage.deserialize("$prefix <red>Failed to add player to the blacklist. Maybe the name you've entered is not correct?</red>"))
                                    return@CommandExecutor
                                } else {
                                    redisController.setHashField("rv-players-blacklist", playerUUID, redisController.getHashField("rv-players-ip", playerUUID) ?: "unknown-ip")
                                    sender.sendMessage(miniMessage.deserialize("$prefix <gray>Added player <aqua>$playerName</aqua> to the blacklist.</gray>"))
                                }
                            }),
                        CommandAPICommand("remove")
                            .withArguments(StringArgument("proxyBlacklistAdd").replaceSuggestions(ArgumentSuggestions.stringCollection {
                                redisController.getHashValuesAsPair("rv-players-blacklist").map { (uuid) -> mojangUtils.getName(UUID.fromString(uuid)) ?: "unknown-name" }
                            }))
                            .withPermission("redivelocity.admin.proxy.blacklist.remove")
                            .executes(CommandExecutor { sender, args ->
                                val playerName = args[0] as String
                                val playerUUID = mojangUtils.getUUID(playerName).toString()
                                redisController.deleteHashField("rv-players-blacklist", playerUUID)
                                sender.sendMessage(miniMessage.deserialize("$prefix <gray>Removed player <aqua>$playerName</aqua> from the blacklist.</gray>"))
                            }),
                        CommandAPICommand("list")
                            .withPermission("redivelocity.admin.proxy.blacklist.list")
                            .executes(CommandExecutor { sender, _ ->
                                val blacklistedPlayers = redisController.getHashValuesAsPair("rv-players-blacklist")
                                val blacklistedPlayersPrettyNames: MutableList<String> = mutableListOf()
                                blacklistedPlayers.forEach { (playerUUID, ip) ->
                                    val playerName = mojangUtils.getName(UUID.fromString(playerUUID)) ?: "unknown-name"
                                    blacklistedPlayersPrettyNames.add("$prefix <aqua><hover:show_text:'<aqua>$playerUUID</aqua>'><click:copy_to_clipboard:$playerUUID>$playerName</click></hover></aqua> <dark_grey>(<aqua>IP: <hover:show_text:'<aqua>Click to copy ip</aqua>'><click:copy_to_clipboard:$ip>$ip</click></hover></aqua>)</dark_grey>")
                                }
                                val blacklistedPlayersPrettyString = blacklistedPlayersPrettyNames.joinToString(separator = "<br>")
                                if (blacklistedPlayers.isEmpty()) {
                                    sender.sendMessage(miniMessage.deserialize("$prefix <gray>There are currently no blacklisted players.</gray>"))
                                } else {
                                    sender.sendMessage(miniMessage.deserialize("$prefix <gray>Currently blacklisted players:<br>$blacklistedPlayersPrettyString</gray>"))
                                }
                            })
                    )
            )
            .register()
    }
}